---  
title: "Quick guide to using factR"  
author: "Fursham Hamid"  
date: "`r Sys.Date()`"  
output:  
  rmdformats::readthedown:  
    highlight: kate  
    number_sections: true  
    toc_depth: 3  
vignette: >  
  %\VignetteIndexEntry{factR}  
  %\VignetteEngine{knitr::rmarkdown}  
---  
  
```{r setup, echo=FALSE, cache=FALSE}  
## Global options  
options(max.print="100")  
knitr::opts_chunk$set(  
  collapse = TRUE,  
  comment = "#>")  
knitr::opts_knit$set(width=75)  
```  
  
  
```{r silentload, include=FALSE}  
#Silent load all dependencies for vignette  
library(factR)  
library(AnnotationHub)  
library(Biostrings)  
library(BSgenome.Mmusculus.UCSC.mm10)  
library(GenomicFeatures)  
library(rtracklayer)  
library(tidyverse) 
```  
  
Here we show the most basic steps for processing custom transcriptome and
predicting novel transcripts for functional protein domains and for susceptibility
to nonsense-mediated decay. We will demonstrate this workflow using a sample GTF 
file that is attached with factR, but you may use your own GTF file to test
this pipeline out. Some steps in factR workflow require inputs from reference 
annotation and genomic sequences, and the example below assumes that you have
imported these files as `ref.gtf` and `sequence` variables respectively. Refer
to the [full description of our workflow](https://fursham-h.github.io/factR/articles/factR.html#materials-needed) f
or details on obtaining these files.

```{r}
# load factR
library(factR)

# import custom transcriptome
gtf <- system.file("extdata", "sc_merged_sample.gtf.gz", package = "factR")
custom.gtf <- importGTF(gtf) 

# updating gene info
custom.gtf <- matchGeneInfo(custom.gtf, ref.gtf)

# shortlisting novel transcripts
custom_new.gtf <- subsetNewTranscripts(custom.gtf, ref.gtf) 

# build cds
custom_new.gtf <- buildCDS(custom_new.gtf, ref.gtf, sequence)

# perform prediction
predictDomains(custom_new.gtf, sequence)
predictNMD(custom_new.gtf)
```


 
 