% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/addremoveexons.R
\name{Add/remove-exons}
\alias{Add/remove-exons}
\alias{addExonstoTx}
\alias{removeExonsfromTx}
\title{Generate hypothetical transcripts by adding/removing exons}
\usage{
addExonstoTx(x, y, drop.unmodified = FALSE, allow.external.exons = FALSE)

removeExonsfromTx(x, y, drop.unmodified = FALSE, ignore.strand = FALSE)
}
\arguments{
\item{x}{GRangesList containing exon coordinates of each transcript to be modified}

\item{y}{GRanges object containing exon coordinates to be added/removed in parallel
to transcripts in x. Length of y and x have to be equal. In addition, x-y pairs
have to be on the same chromosome and strand.}

\item{drop.unmodified}{Whether to drop hypothetical transcripts which were not successfully modified.
(Default; FALSE)}

\item{allow.external.exons}{Whether to allow insertion of exons that fall outside the range of the transcript
(Default: FALSE). We recommend carefully inspecting the hypothetical transcripts 
when switching this argument to TRUE, as some exons may not originate from the
same gene.}
}
\value{
GRangesList containing modified transcripts in x. If drop.unmodified is FALSE,
output object is of same length as x and and contain metadata `modified` to
annotate whether insertion/removal is successful. All other metadata in x will 
not be retained


}
\description{
Adding or removing alternative segments to existing transcripts allow 
functional investigation of hypothetical RNAs. The functions below accepts
two vectors of the same length: x being a GRangesList containing exon coordinates
for each transcript to be modified, and y being a GRanges object containing
exon coordinates to add/remove to/from transcripts in x. The operations are done
in parallel, i.e. first exon in y will be added/remove from the first transcript
in x.
}
